### PR TITLE
Verify CA is false in CSR extension

### PR DESCRIPTION
**Pull request by jaellio**
**Please provide a description of this PR:**

When istiod is configured as the RA, verify the CA field in the BasicConstraint extension of the CSR is set to false. Otherwise, reject CSR.

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [x] Security

### CODE REVIEW

File: security/pkg/pki/ra/common.go
Comment (costinm) [inline security/pkg/pki/ra/common.go:111]: I think this is too permissive and also risky.

We create the CSR in ztunnel or istio-agent (or istiod for dns certs - but those don't need validation since istiod creates them).

There is no point in accepting certs that have extensions unless we decide to create such CSRs. We don't have a requirement to support arbitrary CSRs. 

I don't think ztunnel or agent add extensions - so just drop anything that has an extension.

**Review by costinm — COMMENTED** [review]
Maybe I missed it - where do we check the Subject? 

My suggestion is to check the agent and ztunnel generation code, make sure they do the same thing - and generate a CSR here with the same code. Then compare the one we generate to the one we received. 

The DER is deterministic - we could reuse the same code as agent, really just need to verify ztunnel has same fields. If ztunnel is already divergent - we can check for it too.

This is safest and strictest check - we only approve the CSRs with fields we know we generate, nothing else.

**Comment by jaellio** [comment]
> Maybe I missed it - where do we check the Subject? 

I never claimed this PR verifies the subject? To clarify, by "subject" are you referring to the DN, CN, O, etc?

> My suggestion is to check the agent and ztunnel generation code, make sure they do the same thing - and generate a CSR here with the same code. Then compare the one we generate to the one we received. 

- [ ] Could you please clarify if you are referring to the Istio CSR? Are you suggesting that we generate an Istio CSR on both the client and the server (once the request is received but the server?) and then compare the generated fields? This PR was scoped to verify the extensions in the workflow for Istio configured with an external CA. Rather than specifically checking for the CA field we could reject all istio CSRs with unknown fields.

> The DER is deterministic - we could reuse the same code as agent, really just need to verify ztunnel has same fields. If ztunnel is already divergent - we can check for it too.

I'll need to check that we have access to all of the information we need to recreate the CSR on the server without examining the CSR from the client.

> This is safest and strictest check - we only approve the CSRs with fields we know we generate, nothing else.

**Comment by costinm** [comment]
On Fri, Jul 5, 2024, 20:47 Jackie Elliott ***@***.***> wrote:

> Maybe I missed it - where do we check the Subject?
>
> I never claimed this PR verifies the subject? To clarify, by "subject" are
> you referring to the DN, CN, O, etc?
>
Yes, we shouldn't approve CSRs with arbitrary subject that we can't verify.



My suggestion is to check the agent and ztunnel generation code, make sure
> they do the same thing - and generate a CSR here with the same code. Then
> compare the one we generate to the one we received.
>
> Could you please clarify if you are referring to the Istio CSR? Are you
> suggesting that we generate an Istio CSR on both the client and the server
> (once the request is received but the server?) and then compare the
> generated fields? This PR was scoped to verify the extensions in the
> workflow for Istio configured with an external CA. Rather than specifically
> checking for the CA field we could reject all istio CSRs with unknown
> fields.
>

Yes. We can't compare the full CSR - can't sign - but the idea is to make
sure the CSR template is not including any fields we can't verify and the
agent wouldn't generate.

Having a negative check - this PR verify isCA is not set, another PR to
verify the CN - is not very safe, who knows what extensions or fields get
added to a CSR that have some use - and we blindly approve.

What we really want is to only approve the minimum that agent generates -
and keep parity with the Istio-CA, where we only extract the public key and
SAN and create our clean tempate ( not using anything else in the client
CSR). We can't do this - since we don't have the private key and the real
CA will need the signature, but we can serialize what we expect and
compare.

The DER is deterministic - we could reuse the same code as agent, really

File: security/pkg/pki/ra/common.go
Comment (keithmattix) [reply security/pkg/pki/ra/common.go:111]: I worry this is a bit to restrictive: CRLs for example are specified as an extension IIUC; why wouldn't we allow that usage?

**Review by keithmattix — APPROVED** [review]
PR looks good to me

File: security/pkg/pki/ra/common.go
Comment (costinm) [reply security/pkg/pki/ra/common.go:111]: Who would generate CRLs, and how would we validate them ? 

I don't mind supporting CRLs or any other extension: it needs some design doc to explain how they'll be used, changes in Agent and ZTunnel to add them to the CSR and then changes to validator to allow them. 

Approving CSRs created by some other entities ( not agent/ztunnel) with random features is not supported with Istio CA - which just create a new cert template matching what agent needs and nothing else. 

The problem is that X509 has a lot of options that may have different impact - and for the purpose of Istio we don't need the RA path to have a very different behavior than the normal CA path. 

Certificate validation is supposed to be very restrictive - I don't think many CAs ( ACME or others ) allow random/unexpected stuff in the cert. If we have a feature to support CRL - it should be clearly defined and we should also add it in the normal CA path.

Comment (keithmattix) [reply security/pkg/pki/ra/common.go:111]: > and for the purpose of Istio we don't need the RA path to have a very different behavior than the normal CA path.

Correct me if I'm wrong (I don't spend most of my time with certs), but isn't the whole purpose of having a separate CA vs. RA so that you can have the entity that authenticates the request be different from the entity that actually mints the certificate? If an organization has a requirement that all of their certs have a certain extension set, why should istiod (acting as an RA) care?

**Comment by costinm** [comment]
If an org needs CRL - they most likely should add them as part of the
signer, not have each pod get some CRL somehow ( injection?), pass it in
CSR, have it approved without any check if the CRL is good or not - and
signed.

**Comment by costinm** [comment]
Istiod acting as RA shouldn't care - Istiod acting as an approver does
care, it's the job of approver to make sure the request is correct.

Users can run other approvers as well enforcing other policies - and
verifying.

File: security/pkg/pki/ra/common.go
Comment (bleggett) [reply security/pkg/pki/ra/common.go:111]: IIUC the basic problem is that `istiod` can't actually act as a particularly robust RA, so we have to fence of potential abuses of the CSR model in other ways (like rejecting CSRs that look funny).

**Comment by costinm** [comment]
The basic problem is that K8S signing is not designed for delegated RA  -
the expectation is that the originator of the
request is a component ( pod, user, kubelet ), the identity of the
originator as verified by K8S is included in the request -
and an approver (RA) decide based on the originator, CSR and policies - and
finally the signer signs (and may do additional
checks !).

With Istio - we are both originator and approver, and the information about
originator is lost. It is also quite inefficient -
we create a CSR that we approve with some watchers in the middle. But the
main problem is that the
signer can no longer apply any policy since it doesn't know who the
originator is ( unless it's parsing the SAN and parsing
Istio spiffee perhaps) - they only see Istiod. If the CSR is for a CA -
that's quite reasonable to sign since Istiod may use the
CA and is entitled to ask for one.

Since we use K8S signing this way - as an requester and approver at the
same time -  we must make all the checks that are
no longer possible on the signer.

It would be very clean if Istiod created the private key and CSR - and
returned the private key along with the chain, as it
does for SDS. That would also allow Istiod to handle ztunnel impersonation
more cleanly and add the info it needs ( the
proposed extensions to include the k8s node and pod info for example)

**Comment by keithmattix** [comment]
> Since we use K8S signing this way - as an requester and approver at the
same time -  we must make all the checks that are
no longer possible on the signer.

Ahh I see; this is the part I was missing. Thanks for explaining!

**Comment by jaellio** [comment]
Closing in favor of https://github.com/istio/istio/pull/51966

**Pull request closed without merging**
